Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs

```diff
@@ -161,24 +164,46 @@ protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage
             request.Headers.Authorization = cachedAuthentication;
         }
 
-        var response = await base.SendAsync(request, cancellationToken);
-        if (response is { StatusCode: HttpStatusCode.OK })
-        {
-            return response;
-        }
-        else if (response is { StatusCode: HttpStatusCode.Unauthorized } && TryParseAuthenticationInfo(response, out string? scheme, out AuthInfo? authInfo))
+        int retryCount = 0;
+
+        while (retryCount < MaxRequestRetries)
         {
-            if (await GetAuthenticationAsync(scheme, authInfo.Realm, authInfo.Service, authInfo.Scope, cancellationToken) is AuthenticationHeaderValue authentication)
+            try
             {
-                request.Headers.Authorization = AuthHeaderCache.AddOrUpdate(request.RequestUri, authentication);
-                return await base.SendAsync(request, cancellationToken);
+                var response = await base.SendAsync(request, cancellationToken);
+                if (response is { StatusCode: HttpStatusCode.OK })
+                {
+                    return response;
+                }
+                else if (response is { StatusCode: HttpStatusCode.Unauthorized } && TryParseAuthenticationInfo(response, out string? scheme, out AuthInfo? authInfo))
+                {
+                    if (await GetAuthenticationAsync(scheme, authInfo.Realm, authInfo.Service, authInfo.Scope, cancellationToken) is AuthenticationHeaderValue authentication)
+                    {
+                        request.Headers.Authorization = AuthHeaderCache.AddOrUpdate(request.RequestUri, authentication);
+                        return await base.SendAsync(request, cancellationToken);
+                    }
+                    return response;
+                }
+                else
+                {
+                    return response;
+                }
+            }
+            catch (HttpRequestException e) when (e.InnerException is IOException ioe && ioe.InnerException is SocketException se)
+            {
+                retryCount += 1;
+
+                // TODO: log in a way that is MSBuild-friendly
+                Console.WriteLine($"Encountered a SocketException with message \"{se.Message}\". Pausing before retry.");
+
+                await Task.Delay(TimeSpan.FromSeconds(1.0 * Math.Pow(2, retryCount)));
+
+                // retry
+                continue;
             }
-            return response;
-        }
-        else
-        {
-            return response;
         }
+
+        throw new ApplicationException("Too many retries, stopping");
```

Comment:
how does this look in a stack trace when it flows up? is this an opportunity to have a better message than the current spew?

Reply:
Still not great. Here's a `containerize.exe` run with retries shortened to 1:

```
Containerize: Uploading layer sha256:a603fa5e3b4127f210503aaa6189abf6286ee5a73deeaab460f8f33ebc6b64e2 to ghcr.io
Containerize: Uploading layer sha256:478909de3dddf82d41bf49336ca75e99dda4b994ff95f8b3c7f9929eccf5bd9c to ghcr.io
Containerize: Uploading layer sha256:42b0769710d8d902794c709bb43daa609dbfbcabec0035151652de886592e7f5 to ghcr.io
Containerize: Uploading layer sha256:165f2f6a3267d0655553c5e46e5da98fee9656d18a69d6658864a9ddbfceb716 to ghcr.io
Containerize: Uploading layer sha256:e573684406723b55606ecce5f188244d5ca85ba1fcd17f4394580cb4b4d3bce6 to ghcr.io
Containerize: Uploading layer sha256:144717ed1d2ad9adf0b70f51c506b18e3870d3c8cfb3f4766c863fcdb53823cb to ghcr.io
Containerize: Layer sha256:165f2f6a3267d0655553c5e46e5da98fee9656d18a69d6658864a9ddbfceb716 already existed
Encountered a SocketException with message "An existing connection was forcibly closed by the remote host.". Pausing before retry.
Encountered a SocketException with message "An existing connection was forcibly closed by the remote host.". Pausing before retry.
Encountered a SocketException with message "An existing connection was forcibly closed by the remote host.". Pausing before retry.
Encountered a SocketException with message "An existing connection was forcibly closed by the remote host.". Pausing before retry.
Containerize: Finished uploading layer sha256:144717ed1d2ad9adf0b70f51c506b18e3870d3c8cfb3f4766c863fcdb53823cb to ghcr.io
Containerize: error CONTAINER001: Failed to push to output registry: System.AggregateException: One or more errors occurred. (Too many retries, stopping)
 ---> System.ApplicationException: Too many retries, stopping
   at Microsoft.NET.Build.Containers.AuthHandshakeMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in S:\System.Containers\Microsoft.NET.Build.Containers\AuthHandshakeMessageHandler.cs:line 206
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Microsoft.NET.Build.Containers.Registry.UploadBlob(String name, String digest, Stream contents) in S:\System.Containers\Microsoft.NET.Build.Containers\Registry.cs:line 141
   at Microsoft.NET.Build.Containers.Registry.Push(Layer layer, String name, Action`1 logProgressMessage) in S:\System.Containers\Microsoft.NET.Build.Containers\Registry.cs:line 100
   at Microsoft.NET.Build.Containers.Registry.<>c__DisplayClass18_0.<<Push>b__0>d.MoveNext() in S:\System.Containers\Microsoft.NET.Build.Containers\Registry.cs:line 215
--- End of stack trace from previous location ---
   at Microsoft.NET.Build.Containers.Registry.Push(Image x, String name, String tag, String baseName, Action`1 logProgressMessage) in S:\System.Containers\Microsoft.NET.Build.Containers\Registry.cs:line 191
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Microsoft.NET.Build.Containers.ContainerBuilder.Containerize(DirectoryInfo folder, String workingDir, String registryName, String baseName, String baseTag, String[] entrypoint, String[] entrypointArgs, String imageName, String[] imageTags, String outputRegistry, String[] labels, Port[] exposedPorts, String[] envVars) in S:\System.Containers\Microsoft.NET.Build.Containers\ContainerBuilder.cs:line 75
```

Reply:
thanks - this is fairly clear. I was worried that the 'too many retries' error would mask what else was happening, but we still get incremental updates from the retries prior so my worry was unfounded.